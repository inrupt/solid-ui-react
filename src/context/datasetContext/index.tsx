//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import React, { createContext, ReactElement, useState, useEffect } from "react";
import {
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import useDataset from "../../hooks/useDataset";

export interface IDatasetContext {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo) | undefined;
  setDataset: (solidDataset: SolidDataset) => void;
}

const DatasetContext = createContext<IDatasetContext>({
  solidDataset: undefined,
  setDataset: () => {},
});

export default DatasetContext;

export interface IDatasetProvider {
  children: React.ReactNode;
  /**
   * @deprecated `loading` is deprecated. It can still be used but may be removed in a future major release.`loadingComponent` should be used instead.
   */
  loading?: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
  solidDataset?: SolidDataset | (SolidDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireDatasetOrDatasetUrl =
  | RequireProperty<IDatasetProvider, "solidDataset">
  | RequireProperty<IDatasetProvider, "datasetUrl">;

/**
 * Used to provide a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) to child components through context, as used by various provided components and the useDataset hook.
 */
export const DatasetProvider = ({
  children,
  onError,
  solidDataset: propDataset,
  datasetUrl,
  loading,
  loadingComponent: LoadingComponent,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const { dataset, error } = useDataset(datasetUrl);

  if (error && onError) {
    onError(error);
  }

  const datasetToUse = propDataset ?? dataset;

  // Provide a setDataset function so that child components can update.
  const [stateDataset, setDataset] = useState<
    SolidDataset | (SolidDataset & WithResourceInfo) | undefined
  >(datasetToUse);

  // If the dataset is asynchronously loaded, make sure to set the new state value.
  useEffect(() => {
    setDataset(datasetToUse);
  }, [datasetToUse]);

  let loader: JSX.Element | React.ReactNode | null = (LoadingComponent && (
    <LoadingComponent />
  )) ||
    loading || <span>Fetching data...</span>;

  if (LoadingComponent === null) {
    loader = null;
  }

  return (
    <DatasetContext.Provider value={{ solidDataset: stateDataset, setDataset }}>
      {stateDataset ? children : loader}
    </DatasetContext.Provider>
  );
};
