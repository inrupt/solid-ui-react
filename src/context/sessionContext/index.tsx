/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, {
  createContext,
  ReactElement,
  useState,
  SetStateAction,
  Dispatch,
  useEffect,
  ReactNode,
} from "react";

import {
  fetch,
  login,
  logout,
  handleIncomingRedirect,
  Session,
  getDefaultSession,
  onSessionRestore as onSessionRestoreClient,
} from "@inrupt/solid-client-authn-browser";

import { ILoginInputOptions } from "@inrupt/solid-client-authn-core";

export interface ISessionContext {
  login: (options: ILoginInputOptions) => Promise<void>;
  logout: () => Promise<void>;
  session: Session;
  sessionRequestInProgress: boolean;
  setSessionRequestInProgress?: Dispatch<SetStateAction<boolean>> | any;
  fetch: typeof window.fetch;
}

export const SessionContext = createContext<ISessionContext>({
  login,
  logout,
  fetch,
  session: getDefaultSession(),
  sessionRequestInProgress: true,
});

/* eslint react/require-default-props: 0 */
export interface ISessionProvider {
  children: ReactNode;
  sessionId?: string;
  session?: Session;
  sessionRequestInProgress?: boolean;
  onError?: (error: Error) => void;
  restorePreviousSession?: boolean;
  onSessionRestore?: (url: string) => void;
}

/**
 * Used to provide session data to child components through context, as used by various provided components and the useSession hook.
 */
export const SessionProvider = ({
  sessionId,
  children,
  onError,
  sessionRequestInProgress: defaultSessionRequestInProgress,
  restorePreviousSession,
  onSessionRestore,
}: ISessionProvider): ReactElement => {
  const [session, setSession] = useState<Session>(getDefaultSession());

  if (onSessionRestore !== undefined) {
    onSessionRestoreClient(onSessionRestore);
  }

  const defaultInProgress =
    typeof defaultSessionRequestInProgress === "undefined"
      ? !session.info.isLoggedIn
      : defaultSessionRequestInProgress;

  // If loggedin is true, we're not making a session request.
  const [sessionRequestInProgress, setSessionRequestInProgress] = useState(
    defaultInProgress
  );

  let currentLocation;

  if (typeof window !== "undefined") {
    currentLocation = window.location;
  }

  useEffect(() => {
    handleIncomingRedirect({
      url: window.location.href,
      restorePreviousSession,
    })
      .catch((error: Error) => {
        if (onError) {
          onError(error);
        } else {
          throw error;
        }
      })
      .finally(() => {
        // console.log("done");
        setSessionRequestInProgress(false);
      });

    getDefaultSession().on("logout", () => {
      // TODO force a refresh
      setSession(getDefaultSession());
    });
  }, [session, sessionId, onError, currentLocation, restorePreviousSession]);

  const contextLogin = async (options: ILoginInputOptions) => {
    setSessionRequestInProgress(true);

    try {
      await login(options);
    } catch (error) {
      if (onError) {
        onError(error);
      } else {
        throw error;
      }
    } finally {
      setSessionRequestInProgress(false);
    }
  };

  const contextLogout = async () => {
    try {
      await logout();
    } catch (error) {
      if (onError) {
        onError(error);
      } else {
        throw error;
      }
    }
  };

  return (
    <SessionContext.Provider
      value={{
        session,
        login: contextLogin,
        logout: contextLogout,
        sessionRequestInProgress,
        setSessionRequestInProgress,
        fetch,
      }}
    >
      {children}
    </SessionContext.Provider>
  );
};
