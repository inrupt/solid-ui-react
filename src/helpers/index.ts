//
// Copyright 2022 Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import {
  overwriteFile as solidOverwriteFile,
  getFile,
  Thing,
  SolidDataset,
  UrlString,
  Url,
  getBoolean,
  getDatetime,
  getDecimal,
  getInteger,
  getUrl,
  getStringNoLocale,
  getStringWithLocale,
  getBooleanAll,
  getDatetimeAll,
  getDecimalAll,
  getIntegerAll,
  getStringNoLocaleAll,
  getStringWithLocaleAll,
  getUrlAll,
  getSolidDataset,
} from "@inrupt/solid-client";

import { useContext, useEffect, useState } from "react";

import ThingContext from "../context/thingContext";
import DatasetContext from "../context/datasetContext";

export type CommonProperties = {
  thing?: Thing;
  solidDataset?: SolidDataset;
  property?: Url | UrlString;
  properties?: Url[] | UrlString[];
  edit?: boolean;
  autosave?: boolean;
  onSave?: (savedDataset?: SolidDataset, savedThing?: Thing) => void;
  onError?: (error: Error) => void;
};

export const overwriteFile = async (
  src: string,
  file: File,
  input: EventTarget & HTMLInputElement,
  fetch: typeof window.fetch,
  maxSize?: number,
  onSave?: () => void,
  onError?: (error: Error) => void
): Promise<string | null> => {
  try {
    if (maxSize !== undefined && file.size > maxSize * 1024) {
      input.setCustomValidity(
        `The selected file must not be larger than ${maxSize}kB`
      );
      input.reportValidity();
      return null;
    }
    input.setCustomValidity("");
    await solidOverwriteFile(src, file, { fetch });
    if (onSave) {
      onSave();
    }
    const objectUrl = URL.createObjectURL(file);
    return objectUrl;
  } catch (error) {
    if (onError) {
      onError(error as Error);
    }
    return null;
  }
};

export const retrieveFile = async (
  src: string,
  fetch: typeof window.fetch
): Promise<Blob> => {
  const imageBlob = await getFile(src, { fetch });
  return imageBlob;
};

export type DataType =
  | "boolean"
  | "datetime"
  | "decimal"
  | "integer"
  | "string"
  | "url";

export function getValueByType(
  type: "boolean",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getBoolean>;

export function getValueByType(
  type: "datetime",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getDatetime>;

export function getValueByType(
  type: "decimal",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getDecimal>;

export function getValueByType(
  type: "integer",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getInteger>;

export function getValueByType(
  type: "string",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getStringNoLocale>;

export function getValueByType(
  type: "string",
  thing: Thing,
  property: UrlString | Url,
  locale: string
): ReturnType<typeof getStringWithLocale>;

export function getValueByType(
  type: "string",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getStringNoLocale>;

export function getValueByType(
  type: "url",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getUrl>;

export function getValueByType(
  type: DataType,
  thing: Thing,
  property: UrlString | Url,
  locale?: string
): string | boolean | number | Date | null;

export function getValueByType(
  type: DataType,
  thing: Thing,
  property: UrlString | Url,
  locale?: string
): string | boolean | number | Date | null {
  switch (type) {
    case "boolean":
      return getBoolean(thing, property);
    case "datetime":
      return getDatetime(thing, property);
    case "decimal":
      return getDecimal(thing, property);
    case "integer":
      return getInteger(thing, property);
    case "url":
      return getUrl(thing, property);
    default:
      if (locale) {
        return getStringWithLocale(thing, property, locale);
      }
      return getStringNoLocale(thing, property);
  }
}

export function getValueByTypeAll(
  type: "boolean",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getBooleanAll>;

export function getValueByTypeAll(
  type: "datetime",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getDatetimeAll>;

export function getValueByTypeAll(
  type: "decimal",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getDecimalAll>;

export function getValueByTypeAll(
  type: "integer",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getIntegerAll>;

export function getValueByTypeAll(
  type: "string",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getStringNoLocaleAll>;

export function getValueByTypeAll(
  type: "string",
  thing: Thing,
  property: UrlString | Url,
  locale: string
): ReturnType<typeof getStringWithLocaleAll>;

export function getValueByTypeAll(
  type: "string",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getStringNoLocaleAll>;

export function getValueByTypeAll(
  type: "url",
  thing: Thing,
  property: UrlString | Url
): ReturnType<typeof getUrlAll>;

export function getValueByTypeAll(
  type: DataType,
  thing: Thing,
  property: UrlString | Url,
  locale?: string
): Array<string | boolean | number | Date | null>;

export function getValueByTypeAll(
  type: DataType,
  thing: Thing,
  property: UrlString | Url,
  locale?: string
): Array<string | boolean | number | Date | null> {
  switch (type) {
    case "boolean":
      return getBooleanAll(thing, property);
    case "datetime":
      return getDatetimeAll(thing, property);
    case "decimal":
      return getDecimalAll(thing, property);
    case "integer":
      return getIntegerAll(thing, property);
    case "url":
      return getUrlAll(thing, property);
    default:
      if (locale) {
        return getStringWithLocaleAll(thing, property, locale);
      }
      return getStringNoLocaleAll(thing, property);
  }
}

export function getPropertyForThing(
  propertySelector: typeof getValueByType | typeof getValueByTypeAll,
  type: DataType,
  thing: Thing,
  properties: Array<Url | UrlString>,
  locale?: string
): Url | UrlString | undefined {
  return (
    properties.find((property: Url | UrlString) => {
      return propertySelector(type, thing, property, locale);
    }) || properties[0]
  );
}

export type UseProperty = {
  dataset?: SolidDataset;
  thing?: Thing;
  error?: Error;
  property: Url | UrlString;
  value: string | boolean | number | Date | null;
  setDataset: (dataset: SolidDataset) => void;
  setThing: (thing: Thing) => void;
};

export type UsePropertyProps = {
  thing?: Thing;
  dataset?: SolidDataset;
  property?: Url | UrlString;
  properties?: Url[] | UrlString[];
  type: DataType;
  locale?: string;
};

export function useProperty(props: UsePropertyProps): UseProperty {
  const {
    dataset: propDataset,
    thing: propThing,
    properties: propProperties,
    property: propProperty,
    type,
    locale,
  } = props;

  const { solidDataset: contextDataset, setDataset = () => {} } =
    useContext(DatasetContext);

  const dataset = propDataset || contextDataset;

  const { thing: contextThing, setThing = () => {} } = useContext(ThingContext);

  const thing = propThing || contextThing || undefined;
  let error;

  if (!thing) {
    error = new Error("Thing not found as property or in context");
  }

  const property =
    thing && propProperties
      ? getPropertyForThing(
          getValueByType,
          type,
          thing,
          propProperties,
          locale
        ) || propProperties[0]
      : propProperty;

  if (!property) {
    throw new Error("No property provided");
  }

  const value =
    thing && property ? getValueByType(type, thing, property, locale) : null;

  return {
    dataset,
    thing,
    property,
    error,
    value,
    setDataset,
    setThing,
  };
}

export function useDatetimeBrowserSupport(): boolean | null {
  const [isDatetimeSupported, setIsDatetimeSupported] = useState<
    boolean | null
  >(null);

  useEffect(() => {
    const test = document.createElement("input");
    test.type = "datetime-local";
    setIsDatetimeSupported(test.type !== "text");
  }, []);

  return isDatetimeSupported;
}

export async function updateDataset(
  datasetUrl: string | Url,
  setDataset: (dataset: SolidDataset) => void
): Promise<void> {
  const latestDataset = await getSolidDataset(datasetUrl);
  setDataset(latestDataset);
}
