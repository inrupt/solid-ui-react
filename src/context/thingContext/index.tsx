/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, {
  createContext,
  ReactElement,
  useState,
  useEffect,
  useContext,
} from "react";
import {
  Thing,
  LitDataset,
  WithResourceInfo,
  UrlString,
  getThingOne,
} from "@inrupt/solid-client";
import { DatasetContext } from "../datasetContext";

interface IThingContext {
  thing: Thing | undefined;
}

export const ThingContext = createContext<IThingContext>({
  thing: undefined,
});

interface IThingProvider {
  children: React.ReactNode;
  thing?: Thing;
  thingUrl?: UrlString | string;
  dataset?: LitDataset | (LitDataset & WithResourceInfo);
  datasetUrl?: UrlString | string;
}

type RequireProperty<T, Prop extends keyof T> = T & { [key in Prop]-?: T[key] };
type RequireDatasetOrDatasetUrl =
  | RequireProperty<IThingProvider, "thing">
  | RequireProperty<IThingProvider, "thingUrl">;

export const ThingProvider = ({
  children,
  thing,
  thingUrl,
}: RequireDatasetOrDatasetUrl): ReactElement => {
  const datasetContext = useContext(DatasetContext);
  const { dataset } = datasetContext;
  const [litThing, setLitThing] = useState<Thing>();

  useEffect(() => {
    if (dataset && thingUrl) {
      setLitThing(getThingOne(dataset, thingUrl));
    }
  }, [dataset, thingUrl]);

  if (thing) {
    return (
      <ThingContext.Provider
        value={{
          thing,
        }}
      >
        {children}
      </ThingContext.Provider>
    );
  }

  return (
    <ThingContext.Provider
      value={{
        thing: litThing,
      }}
    >
      {children}
    </ThingContext.Provider>
  );
};
