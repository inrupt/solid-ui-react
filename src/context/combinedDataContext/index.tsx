//
// Copyright Inrupt Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to use,
// copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
// Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
// SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
//

import type { ReactElement } from "react";
import React from "react";

import type {
  Thing,
  SolidDataset,
  WithResourceInfo,
  UrlString,
} from "@inrupt/solid-client";

import { DatasetProvider } from "../datasetContext";
import { ThingProvider } from "../thingContext";

export type Props = {
  children: React.ReactNode;
  loadingComponent?: React.ComponentType | null;
  onError?(error: Error): void | null;
};

export type ThingOrThingUrl = {
  thing?: Thing;
  thingUrl?: UrlString | string;
};

export interface Dataset extends Props {
  solidDataset: SolidDataset | (SolidDataset & WithResourceInfo);
}

export interface DatasetUrl extends Props {
  datasetUrl: UrlString | string;
}

export type RequireProperty<T, Prop extends keyof T> = T & {
  [key in Prop]-?: T[key];
};

export type RequireThingOrThingUrl =
  | RequireProperty<ThingOrThingUrl, "thing">
  | RequireProperty<ThingOrThingUrl, "thingUrl">;

// TODO: naming of this provider TBC
function CombinedDataProvider(
  props: Dataset & RequireThingOrThingUrl,
): ReactElement;

function CombinedDataProvider(
  props: DatasetUrl & RequireThingOrThingUrl,
): ReactElement;

/**
 * Used to provide both a [Dataset](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-SolidDataset) and [Thing](https://docs.inrupt.com/developer-tools/javascript/client-libraries/reference/glossary/#term-Thing) to child components through context, as used by various provided components and the useDataset and useThing hooks.
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
function CombinedDataProvider(props: any): ReactElement {
  const {
    children,
    solidDataset,
    datasetUrl,
    thing,
    thingUrl,
    onError,
    loadingComponent,
  } = props;

  return (
    <DatasetProvider
      onError={onError}
      solidDataset={solidDataset}
      datasetUrl={datasetUrl}
      loadingComponent={loadingComponent}
    >
      <ThingProvider thing={thing} thingUrl={thingUrl}>
        {children}
      </ThingProvider>
    </DatasetProvider>
  );
}

export default CombinedDataProvider;
