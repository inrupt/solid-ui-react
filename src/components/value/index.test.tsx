/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

/* eslint-disable @typescript-eslint/no-explicit-any */
import * as React from "react";
import { render } from "@testing-library/react";
import * as SolidFns from "@inrupt/solid-client";
import { Value } from "./index";
import * as helpers from "../../helpers";
import { DatasetProvider } from "../../context/datasetContext";
import { ThingProvider } from "../../context/thingContext";

const mockPredicate = `http://xmlns.com/foaf/0.1/nick`;
const mockNick = "test nick value";

const mockBdayPredicate = "http://www.w3.org/2006/vcard/ns#bday";
const mockBday = new Date("2021-05-04T06:00:00.000Z");

const mockThing = SolidFns.addStringNoLocale(
  SolidFns.createThing(),
  mockPredicate,
  mockNick
);

const mockThingWithBday = SolidFns.addDatetime(
  mockThing,
  mockBdayPredicate,
  mockBday
);

const mockDataset = SolidFns.setThing(
  SolidFns.mockSolidDatasetFrom("https://some-interesting-value.com"),
  mockThing
);
const mockDatasetWithResourceInfo = SolidFns.setThing(
  SolidFns.createSolidDataset() as any,
  mockThing
);

const inputOptions = {
  name: "test-name",
  type: "url",
};

const mockDatasetWithBdayResourceInfo = SolidFns.setThing(
  SolidFns.createSolidDataset(),
  mockThingWithBday
);

const savedDataset = SolidFns.setThing(
  SolidFns.mockSolidDatasetFrom("https://example.pod/resource"),
  SolidFns.createThing()
);
jest.spyOn(SolidFns, "saveSolidDatasetAt").mockResolvedValue(savedDataset);

describe("<Value /> component snapshot test", () => {
  it("matches snapshot", () => {
    const documentBody = render(
      <Value
        dataType="string"
        solidDataset={mockDatasetWithResourceInfo}
        thing={mockThing}
        property={mockPredicate}
      />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("matches snapshot while fetching data", () => {
    jest.spyOn(helpers, "useProperty").mockReturnValueOnce({
      thing: undefined,
      error: undefined,
      value: null,
      setDataset: jest.fn(),
      setThing: jest.fn(),
      property: mockPredicate,
    });
    const documentBody = render(
      <Value dataType="string" property={mockPredicate} />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders loading component if passed while fetching data", () => {
    jest.spyOn(helpers, "useProperty").mockReturnValueOnce({
      thing: undefined,
      error: undefined,
      value: null,
      setDataset: jest.fn(),
      setThing: jest.fn(),
      property: mockPredicate,
    });
    const documentBody = render(
      <Value
        dataType="string"
        property={mockPredicate}
        loadingComponent={() => (
          <span id="custom-loading-component">loading...</span>
        )}
      />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders default error message if there is an error", () => {
    const emptyThing = SolidFns.createThing();
    const documentBody = render(
      <Value
        thing={emptyThing}
        dataType="string"
        property="https://example.com/bad-url"
      />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders custom error component if passed and there is an error", () => {
    const emptyThing = SolidFns.createThing();
    const documentBody = render(
      <Value
        thing={emptyThing}
        dataType="string"
        property="https://example.com/bad-url"
        errorComponent={({ error }) => (
          <span id="custom-error-component">{error.toString()}</span>
        )}
      />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders 'false' if dataType is boolean instead of error if property not found", () => {
    const emptyThing = SolidFns.createThing();
    const documentBody = render(
      <DatasetProvider solidDataset={mockDatasetWithResourceInfo}>
        <ThingProvider thing={mockThing}>
          <Value
            thing={emptyThing}
            dataType="boolean"
            property="http://xmlns.com/foaf/0.1/name"
          />
        </ThingProvider>
      </DatasetProvider>
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("renders error if dataType is boolean and there is an error fetching the thing", () => {
    const documentBody = render(
      <Value
        thing={undefined}
        dataType="boolean"
        property="http://xmlns.com/foaf/0.1/name"
      />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("matches snapshot with data from context", () => {
    const documentBody = render(
      <DatasetProvider solidDataset={mockDatasetWithResourceInfo}>
        <ThingProvider thing={mockThing}>
          <Value dataType="string" property={mockPredicate} edit autosave />
        </ThingProvider>
      </DatasetProvider>
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("matches snapshot with edit true and inputOptions", () => {
    const documentBody = render(
      <Value
        dataType="string"
        edit
        inputProps={inputOptions}
        solidDataset={mockDatasetWithResourceInfo}
        thing={mockThing}
        property={mockPredicate}
      />
    );
    const { baseElement } = documentBody;
    expect(baseElement).toMatchSnapshot();
  });

  it("matches snapshot with dataType datetime when browser supports datetime-local", () => {
    const { asFragment } = render(
      <Value
        dataType="datetime"
        edit
        solidDataset={mockDatasetWithBdayResourceInfo}
        thing={mockThingWithBday}
        property={mockBdayPredicate}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it.each([["string"], ["boolean"], ["decimal"], ["integer"], ["url"]])(
    "matches snapshot with dataType %i",
    (dataType) => {
      const { asFragment } = render(
        <Value
          dataType={dataType as helpers.DataType}
          edit
          solidDataset={mockDataset}
          thing={mockThing}
          property={mockPredicate}
        />
      );
      expect(asFragment()).toMatchSnapshot();
    }
  );

  it("matches snapshot with dataType datetime when browser does not support datetime-local", () => {
    jest.spyOn(helpers, "useDatetimeBrowserSupport").mockReturnValueOnce(false);
    const { asFragment } = render(
      <Value
        dataType="datetime"
        edit
        solidDataset={mockDatasetWithBdayResourceInfo}
        thing={mockThingWithBday}
        property={mockBdayPredicate}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
