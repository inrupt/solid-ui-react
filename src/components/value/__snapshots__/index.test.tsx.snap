// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Value /> component snapshot test matches snapshot 1`] = `
<body>
  <div>
    <span>
      test nick value
    </span>
  </div>
</body>
`;

exports[`<Value /> component snapshot test matches snapshot with data from context 1`] = `
<body>
  <div>
    <input
      type="text"
      value="test nick value"
    />
  </div>
</body>
`;

exports[`<Value /> component snapshot test matches snapshot with dataType datetime when browser does not support datetime-local 1`] = `
<DocumentFragment>
  <input
    aria-label="Date"
    pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}"
    placeholder="yyyy-mm-dd"
    step="any"
    type="date"
    value="2021-05-04"
  />
  <input
    aria-label="Time"
    pattern="[0-9]{2}:[0-9]{2}"
    type="time"
    value="06:00:00"
  />
</DocumentFragment>
`;

exports[`<Value /> component snapshot test matches snapshot with dataType datetime when browser supports datetime-local 1`] = `
<DocumentFragment>
  <input
    pattern="[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}"
    placeholder="yyyy-mm-ddThh:mm"
    step="any"
    type="datetime-local"
    value="2021-05-04T06:00:00"
  />
</DocumentFragment>
`;

exports[`<Value /> component snapshot test matches snapshot with edit true and inputOptions 1`] = `
<body>
  <div>
    <input
      name="test-name"
      type="url"
      value="test nick value"
    />
  </div>
</body>
`;
